# Imports

In [1]:
%reload_ext autoreload
%autoreload 2

In [2]:
import sys
import pathlib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import plotly.figure_factory as ff
from plotly.subplots import make_subplots

ROOT = pathlib.Path().absolute().parent
RAW_DATA_PATH = ROOT / 'data' / 'raw'
PROCESSED_DATA_PATH = ROOT / 'data' / 'processed'

# endure this project is in the path
sys.path.insert(0, ROOT.absolute().as_posix())
from src.visualization.visualize import SalesVisualizations

# Calendar 

Contains information about the dates the products are sold.
- date: The date in a “y-m-d” format.
- wm_yr_wk: The id of the week the date belongs to.
- weekday: The type of the day (Saturday, Sunday, …, Friday).
- wday: The id of the weekday, starting from Saturday.
- month: The month of the date.
- year: The year of the date.
- event_name_1: If the date includes an event, the name of this event.
- event_type_1: If the date includes an event, the type of this event.
- event_name_2: If the date includes a second event, the name of this event.
- event_type_2: If the date includes a second event, the type of this event.
- snap_CA, snap_TX, and snap_WI: A binary variable (0 or 1) indicating whether the stores of CA,
TX or WI allow SNAP2 purchases on the examined date. 1 indicates that SNAP purchases are
allowed. - SNAP is known colloquially as "Food Stamps"

In [8]:
categorical_variables = [
    'weekday','wday','month',
    'event_name_1','event_type_1',
    'event_name_2','event_type_2']
boolean_variables = ['snap_CA','snap_TX','snap_WI']

dtypes = {var: 'category' for var in categorical_variables}
dtypes['d'] = 'string'

# read the data with chosen data types
df = \
(pd
.read_csv(RAW_DATA_PATH / 'calendar.csv',
          parse_dates=['date'], dtype=dtypes)
.astype({var: 'boolean' for var in boolean_variables}))

df.info(memory_usage='deep')
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1969 entries, 0 to 1968
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date          1969 non-null   datetime64[ns]
 1   wm_yr_wk      1969 non-null   int64         
 2   weekday       1969 non-null   category      
 3   wday          1969 non-null   category      
 4   month         1969 non-null   category      
 5   year          1969 non-null   int64         
 6   d             1969 non-null   string        
 7   event_name_1  162 non-null    category      
 8   event_type_1  162 non-null    category      
 9   event_name_2  5 non-null      category      
 10  event_type_2  5 non-null      category      
 11  snap_CA       1969 non-null   boolean       
 12  snap_TX       1969 non-null   boolean       
 13  snap_WI       1969 non-null   boolean       
dtypes: boolean(3), category(7), datetime64[ns](1), int64(2), string(1)
memory usage: 93.1 KB

date  wm_yr_wk    weekday wday month  year    d event_name_1  \
0 2011-01-29     11101   Saturday    1     1  2011  d_1          NaN   
1 2011-01-30     11101     Sunday    2     1  2011  d_2          NaN   
2 2011-01-31     11101     Monday    3     1  2011  d_3          NaN   
3 2011-02-01     11101    Tuesday    4     2  2011  d_4          NaN   
4 2011-02-02     11101  Wednesday    5     2  2011  d_5          NaN   

  event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          NaN          NaN          NaN    False    False    False  
1          NaN          NaN          NaN    False    False    False  
2          NaN          NaN          NaN    False    False    False  
3          NaN          NaN          NaN     True     True    False  
4          NaN          NaN          NaN     True    False     True

Memory usage reduced approx. 10x


In [9]:
event_names = pd.concat([df.event_name_1,df.event_name_2]).dropna().unique().tolist()
event_types = pd.concat([df.event_type_1,df.event_type_2]).dropna().unique().tolist()

print(f"""There are {len(event_names)} unique events:""")
for row in zip(event_names[:10], event_names[10:20], event_names[20:]):
    print(str('{:<20}' * 3).format(*row))

print(f"""\nThere are {len(event_types)} unique event types:""")
print(*event_types, sep='\n')


There are 30 unique events:
SuperBowl           Mother's day        Halloween           
ValentinesDay       MemorialDay         EidAlAdha           
PresidentsDay       NBAFinalsStart      VeteransDay         
LentStart           NBAFinalsEnd        Thanksgiving        
LentWeek2           Father's day        Christmas           
StPatricksDay       IndependenceDay     Chanukah End        
Purim End           Ramadan starts      NewYear             
OrthodoxEaster      Eid al-Fitr         OrthodoxChristmas   
Pesach End          LaborDay            MartinLutherKingDay 
Cinco De Mayo       ColumbusDay         Easter              

There are 4 unique event types:
Sporting
Cultural
National
Religious


In [10]:
print("Breakdown of event types:")
df.groupby('event_name_1', observed=True).first().groupby('event_type_1').size().sort_values()

Breakdown of event types:


event_type_1
Sporting      3
Cultural      7
National     10
Religious    10
dtype: int64

In [11]:
df[df.event_name_2.notnull()].groupby(
    ['year','month','weekday',
    'event_name_1','event_type_1',
    'event_name_2','event_type_1'], observed=True).size()

year  month  weekday  event_name_1    event_type_1  event_name_2    event_type_1
2011  4      Sunday   OrthodoxEaster  Religious     Easter          Religious       1
2013  5      Sunday   OrthodoxEaster  Religious     Cinco De Mayo   Religious       1
2014  4      Sunday   Easter          Cultural      OrthodoxEaster  Cultural        1
      6      Sunday   NBAFinalsEnd    Sporting      Father's day    Sporting        1
2016  6      Sunday   NBAFinalsEnd    Sporting      Father's day    Sporting        1
dtype: int64

Across the entire 5 year period there are only 5 dates with two events.

In [12]:
display(df.filter(like='snap').sum())

print(
"""Proportion of dates where all states have 'snap' at the same time\n"""
f"""{df.filter(like='snap').all(axis='columns').mean().round(2)}""")

snap_CA    650.0
snap_TX    650.0
snap_WI    650.0
dtype: float64

Proportion of dates where all states have 'snap' at the same time
0.13


Over the 5 year period each state has the same number of 'snap' dates, but they do not fall on the same dates.

## Further Analysis on Calendar

Could look further into the distribution of snap dates throughout the year, how much do the states differ? Could try making a caledar plot.

What happens when a snap date coincides with an event?

Do the events happen every year & how many occur on the same date? Does it matter that Cinco de Mayo can be any day of the week while LentStart is always Wednesday.


# Sell Prices

Contains information about the price of the products sold per store and date.
- store_id: The id of the store where the product is sold.
- item_id: The id of the product.
- wm_yr_wk: The id of the week.
- sell_price: The price of the product for the given week/store. The price is provided per week
(average across seven days). If not available, this means that the product was not sold during the
examined week. Note that although prices are constant at weekly basis, they may change through
time (both training and test set). 

In [13]:
# need to check that all sell_prices are in whole cents
# load one column of the csv as a series
# find the length of the string right of the decimal point
# (pd
# .read_csv(
#     RAW_DATA_PATH / 'sell_prices.csv',
#     usecols=['sell_price'],
#     squeeze=True)
# .astype(str)
# .str.rsplit('.', n=1)
# .str.len()
# .value_counts())

957.5 MB -> 46.0 MB and far quicker to load

In [4]:
df = pd.read_feather(PROCESSED_DATA_PATH / 'sell_prices_processed.feather')

df.info(memory_usage='deep')
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6841121 entries, 0 to 6841120
Data columns (total 7 columns):
 #   Column           Dtype   
---  ------           -----   
 0   state_id         category
 1   store_id         category
 2   cat_id           category
 3   dept_id          category
 4   item_id          category
 5   wm_yr_wk         int16   
 6   sell_price_cent  int16   
dtypes: category(5), int16(2)
memory usage: 65.5 MB


state_id store_id   cat_id    dept_id        item_id  wm_yr_wk  \
0       CA     CA_1  HOBBIES  HOBBIES_1  HOBBIES_1_001     11325   
1       CA     CA_1  HOBBIES  HOBBIES_1  HOBBIES_1_001     11326   
2       CA     CA_1  HOBBIES  HOBBIES_1  HOBBIES_1_001     11327   
3       CA     CA_1  HOBBIES  HOBBIES_1  HOBBIES_1_001     11328   
4       CA     CA_1  HOBBIES  HOBBIES_1  HOBBIES_1_001     11329   

   sell_price_cent  
0              958  
1              958  
2              826  
3              826  
4              826

In [15]:
# import dask.dataframe as dd
# df = dd.read_parquet(PROCESSED_DATA_PATH / 'sell_prices_processed.parquet')
# df.persist().isna().sum().sum().compute()

df.describe(include='all')

state_id store_id   cat_id  dept_id          item_id      wm_yr_wk  \
count   6841121  6841121  6841121  6841121          6841121  6.841121e+06   
unique        3       10        3        7             3049           NaN   
top          CA     TX_2    FOODS  FOODS_3  HOUSEHOLD_1_460           NaN   
freq    2708822   701214  3181789  1835270             2820           NaN   
mean        NaN      NaN      NaN      NaN              NaN  1.138294e+04   
std         NaN      NaN      NaN      NaN              NaN  1.486100e+02   
min         NaN      NaN      NaN      NaN              NaN  1.110100e+04   
25%         NaN      NaN      NaN      NaN              NaN  1.124700e+04   
50%         NaN      NaN      NaN      NaN              NaN  1.141100e+04   
75%         NaN      NaN      NaN      NaN              NaN  1.151700e+04   
max         NaN      NaN      NaN      NaN              NaN  1.162100e+04   

        sell_price_cent  
count      6.841121e+06  
unique              NaN  
top                 NaN  
freq                NaN  
mean       4.410792e+02  
std        3.408868e+02  
min        1.000000e+00  
25%        2.180000e+02  
50%        3.470000e+02  
75%        5.840000e+02  
max        1.073200e+04

There are no missing values, and as expected we see the 10 stores in 3 states, and 3049 products in 7 departments within 3 categories.

In [16]:
print(*df.store_id.cat.categories.to_list(), sep=' | ')

CA_1 | CA_2 | CA_3 | CA_4 | TX_1 | TX_2 | TX_3 | WI_1 | WI_2 | WI_3


So we have 4 stores in CA and 3 each in WI and TX.

In [17]:
df.groupby(['store_id','item_id'], observed=True).first().groupby('store_id').size()

store_id
CA_1    3049
CA_2    3049
CA_3    3049
CA_4    3049
TX_1    3049
TX_2    3049
TX_3    3049
WI_1    3049
WI_2    3049
WI_3    3049
dtype: int64

All stores sell all the products at some time, so different stores haven't chosen to stock different things.

In [5]:
# find the amount of each product type
dep_breakdown = \
(df
.groupby('item_id')
.first()
.groupby(['dept_id'])
.size()
.rename('amount')
.sort_index()
.to_frame()
.reset_index()
.rename(columns={'dept_id':'group'}))

cat_breakdown = \
(df
.groupby('item_id')
.first()
.groupby(['cat_id'])
.size()
.rename('amount')
.sort_index()
.to_frame()
.reset_index()
.rename(columns={'cat_id':'group'}))

breakdown = pd.concat(
    [dep_breakdown,cat_breakdown],
    keys=['departments','categories'],
    names=['supergroup'],
    axis=0).reset_index()

px.bar(breakdown, x='group', y='amount', color='supergroup', title="Items in each Category anbd Department")

There is some imbalance in product count in categories and departments.

In [19]:
# how many of each category are on sale over time

# first get the number count of each product in each state over time (sum of all stores)
to_plot = \
(df
.groupby(['cat_id','state_id','item_id','wm_yr_wk'], observed=True)
.agg(item_count=('sell_price_cent','count'))
.reset_index())

# divide through by the number of stores in that state
for state, n_stores in {'CA':4,'WI':3,'TX':3}.items():
    to_plot.item_count = np.where(
        to_plot.state_id.values==state, # condition
        to_plot.item_count.values/n_stores, # where true 
        to_plot.item_count.values) # where false

# num of items in that cat in that state in that week   
to_plot = \
(to_plot
.groupby(['cat_id','state_id','wm_yr_wk'], observed=True)
.agg(item_count=('item_count','sum'))
.reset_index())

# match wm_yr_wk to a date
cal_df = pd.read_csv(
    RAW_DATA_PATH / 'calendar.csv',
    parse_dates=['date'],
    usecols=['date','wm_yr_wk'])

# join to dates and resample to find the monthly average
to_plot = \
(to_plot
.merge(
    cal_df,
    left_on=['wm_yr_wk'],
    right_on=['wm_yr_wk'])
.groupby(['cat_id','state_id'])
.resample('M', on='date')
.mean()
.reset_index('date'))

# display(to_plot.sample(5))

category_ids = to_plot.index.get_level_values('cat_id').categories.to_list()
state_ids = to_plot.index.get_level_values('state_id').categories.to_list()

fig = make_subplots(rows=1, cols=3, column_titles=category_ids)

# plot as lines
for col, category_id in enumerate(category_ids):
    for state_id in state_ids:
        tmp = to_plot.loc[(category_id, state_id), :]
        fig.add_trace(
            go.Scattergl(
                x=tmp.date,
                y=tmp.item_count,
                name=state_id,
                legendgroup=state_id,
                showlegend=not col,
                mode='lines',
                marker=go.scattergl.Marker(
                    color={'CA':'blue','TX':'red','WI':'green'}[state_id])),
            row=1, col=col+1)

layout = go.Layout(
    width=1200,
    title="Monthly item count by category and state",
    yaxis_title="Total items available")

fig.update_layout(layout)
fig

Over time more products are added in all 3 categories with the 3 states following very similar paths and slowing down at the start of 2016. The greatest difference is in foods, WI starting with the fewest but adding products the fasts, and CA adding products at the slowest rate.

In [21]:
# mean price of all products in a category over time
to_plot = \
(df
.groupby(['cat_id','state_id','wm_yr_wk'], observed=True)
.agg(mean_item_price=('sell_price_cent','mean'))
.reset_index())

# match wm_yr_wk to a date
cal_df = pd.read_csv(
    RAW_DATA_PATH / 'calendar.csv',
    parse_dates=['date'],
    usecols=['date','wm_yr_wk'])

# resample to find the monthly average
to_plot = \
(to_plot
.merge(
    cal_df,
    left_on=['wm_yr_wk'],
    right_on=['wm_yr_wk'])
.groupby(['cat_id','state_id'])
.resample('M', on='date')
.mean_item_price
.mean()
.reset_index('date'))

# display(to_plot.sample(5))


category_ids = to_plot.index.get_level_values('cat_id').categories.to_list()
state_ids = to_plot.index.get_level_values('state_id').categories.to_list()

fig = make_subplots(rows=1, cols=3, column_titles=category_ids)

# plot as lines
for col, category_id in enumerate(category_ids):
    for state_id in state_ids:
        tmp = to_plot.loc[(category_id, state_id), :]
        fig.add_trace(
            go.Scattergl(
                x=tmp.date,
                y=tmp.mean_item_price,
                name=state_id,
                legendgroup=state_id,
                showlegend=not col,
                mode='lines',
                marker=go.scattergl.Marker(
                    color={'CA':'blue','TX':'red','WI':'green'}[state_id])),
            row=1, col=col+1)

layout = go.Layout(
    width=1200,
    title="Monthly mean item price by category and state",
    yaxis_title="price / cent")

fig.update_layout(layout)
display(fig)
display(fig.for_each_xaxis(lambda x: x.update(range=['2015-01-01','2016-07-01'])))

The average prices over all goods has varied a lot over the last 5 years, but with the introduction of so many new products this is to be expected. If we look just at the last 18 months where very few, if any, new products were added, the prices are far more stable, with household items increasing in price a little across the board and foods costs increasing significantly in Texas and California - possibly due to lack of rainfall.

In [22]:
# mean price of an item in a state across stores
to_plot = \
(df
# .sample(frac=0.01)
.groupby(['cat_id','state_id','item_id','wm_yr_wk'], observed=True)
.agg(item_price=('sell_price_cent','mean'))
.reset_index())

# match wm_yr_wk to a date
cal_df = pd.read_csv(
    RAW_DATA_PATH / 'calendar.csv',
    parse_dates=['date'],
    usecols=['date','wm_yr_wk']).groupby('wm_yr_wk').first()

# join in the dates
to_plot = \
(to_plot
.merge(
    cal_df,
    left_on=['wm_yr_wk'],
    right_on=['wm_yr_wk']))

# standardise the item_price
to_plot.item_price = \
(to_plot
[['state_id','item_id','item_price']]
.groupby(['state_id','item_id'], observed=True)
.transform(lambda x: (x-x.mean())/x.std()))

# get monthly mean item price across states and categories
to_plot = \
(to_plot
.groupby(['cat_id','state_id'], observed=True)
.resample('M', on='date')
.item_price
.mean()
.reset_index('date'))

# display(to_plot.sample(5))

category_ids = to_plot.index.get_level_values('cat_id').categories.to_list()
state_ids = to_plot.index.get_level_values('state_id').categories.to_list()

fig = make_subplots(rows=1, cols=3, column_titles=category_ids)

# plot as lines
for col, category_id in enumerate(category_ids):
    for state_id in state_ids:
        tmp = to_plot.loc[(category_id, state_id), :]
        fig.add_trace(
            go.Scattergl(
                x=tmp.date,
                y=tmp.item_price,
                name=state_id,
                legendgroup=state_id,
                showlegend=not col,
                mode='lines',
                marker=go.scattergl.Marker(
                    color={'CA':'blue','TX':'red','WI':'green'}[state_id])),
            row=1, col=col+1)

layout = go.Layout(
    width=1200,
    title="Monthly (Standardized) mean item price by category and state",
    yaxis_title="standardized price")

fig.update_layout(layout)
fig

All three states follow very similar paths with no one state consistently being more or less expoensive than another. The most obvious feature is the large drop in price in November 2014 which coincides with <a href="https://corporate.walmart.com/newsroom/2014/10/06/walmart-announces-new-commitment-to-a-sustainable-food-system-at-global-milestone-meeting">a commitment to sustainability and affordability of food</a>. The price of food rebopunded a few months latewr but has settled at a price far lower than before nov-2014.

Hobbies products were slowly reducing in price until this date but have siunce risen in price quickly.

<strong>THIS STILL DOESN'T ANSWER THE QUESTION. THESE PRICES CAN FLUCTUATE WHEN A BUNCH OF NEW CHEAP/EXPENSIVE ITEMS ARE ADDED TO THE STORE. I WANT TO KNOW, ARE THE PRICES OF THE SAME ITEMS GOING UP OR DOWN???</strong>

In [8]:
# are items miore expensive in different states?
state_category_mean_price = df.copy()#.sample(frac=0.01)

# average price of each product
overall_item_means = state_category_mean_price.groupby(['cat_id','item_id'], observed=True).agg({'sell_price_cent':'mean'})

# % price difference of each product in each state compared to all states
price_diff = state_category_mean_price.groupby(['state_id','cat_id','item_id'], observed=True).agg({'sell_price_cent':'mean'})
price_diff = 100 * (price_diff - overall_item_means) / overall_item_means

# now get the average difference over item categories for each state
price_diff = \
(price_diff
.groupby(['state_id','cat_id'])
.mean()
.rename(columns={'sell_price_cent':'price_diff'}))

state_category_mean_price = \
(state_category_mean_price
.groupby(['state_id','cat_id'], observed=True)
.agg({'sell_price_cent':'mean'})
.join(price_diff)
.reset_index())

px.bar(state_category_mean_price, x='cat_id', y='price_diff', color='state_id', barmode='group')

In [9]:
# mean price of all products in a state over time
to_plot = \
(df
# .sample(frac=0.01)
.groupby(['state_id','wm_yr_wk'], observed=True)
.agg({'sell_price_cent':'mean'})
.reset_index())

# match wm_yr_wk to a date
cal_df = pd.read_csv(
    RAW_DATA_PATH / 'calendar.csv',
    parse_dates=['date'],
    usecols=['date','wm_yr_wk'])

# resample to find the monthly average
to_plot = \
(to_plot
.merge(
    cal_df,
    left_on=['wm_yr_wk'],
    right_on=['wm_yr_wk'])
.groupby('state_id')
.resample('M', on='date')
.mean()
.reset_index('date'))

# display(to_plot.sample(5))

# plot as lines
traces = []
for state_id in to_plot.index.unique():
    tmp = to_plot.loc[state_id, :]
    traces.append(
        go.Scattergl(
            x=tmp.date,
            y=tmp.sell_price_cent,
            name=state_id,
            mode='lines'))

layout = go.Layout(
    title="Monthly average item price by state",
    yaxis_title="Price / cent")

fig = go.Figure(traces, layout)
fig



## Further Work

- have not looked at departments within categories
- are prices of the same items changin over time
- Are prices in different states notably different for the same items
- Are prices different across stores
- are some products seasonal?

# Sales

Contains the historical daily unit sales data per product and store.
- item_id: The id of the product.
- dept_id: The id of the department the product belongs to.
- cat_id: The id of the category the product belongs to.
- store_id: The id of the store where the product is sold.
- state_id: The State where the store is located.
- d_1, d_2, …, d_i, … d_1941: The number of units sold at day i, starting from 2011-01-29.

In [25]:
df = pd.read_feather(PROCESSED_DATA_PATH / 'sales_train_validation_processed.feather')
df.sample(5)

item_id      dept_id     cat_id store_id state_id  d_1  d_2  \
3042       FOODS_3_821      FOODS_3      FOODS     CA_1       CA    0    0   
9056       FOODS_3_736      FOODS_3      FOODS     CA_3       CA    0    2   
11002      FOODS_2_028      FOODS_2      FOODS     CA_4       CA    0    0   
16033  HOUSEHOLD_1_228  HOUSEHOLD_1  HOUSEHOLD     TX_2       TX    0    1   
20271      FOODS_2_151      FOODS_2      FOODS     TX_3       TX    0    0   

       d_3  d_4  d_5  ...  d_1904  d_1905  d_1906  d_1907  d_1908  d_1909  \
3042     0    0    0  ...       1       2       6       3       2       0   
9056     0    1    1  ...       0       0       0       0       0       0   
11002    0    0    0  ...       0       1       1       0       0       1   
16033    1    2    0  ...       0       1       0       0       3       2   
20271    0    0    0  ...       0       0       1       1       1       2   

       d_1910  d_1911  d_1912  d_1913  
3042        1       3       5       1  
9056        1       0       1       1  
11002       0       0       1       0  
16033       3       2       1       0  
20271       0       1       0       2  

[5 rows x 1918 columns]

In [26]:
df.iloc[:, :6].info(memory_usage='deep')
print('\n')
df.iloc[:, 6:].info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   item_id   30490 non-null  object  
 1   dept_id   30490 non-null  category
 2   cat_id    30490 non-null  category
 3   store_id  30490 non-null  category
 4   state_id  30490 non-null  category
 5   d_1       30490 non-null  int16   
dtypes: category(4), int16(1), object(1)
memory usage: 2.2 MB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Columns: 1912 entries, d_2 to d_1913
dtypes: int16(1912)
memory usage: 111.2 MB


Memory reduced 450 MB -> 111 MB

But we will instead use an aggregated dataset which is more convenient to create plots from.

In [8]:
sale_viz = SalesVisualizations()
sale_viz.df.head()

snap       sales  \
                                                    snap   HOUSEHOLD   
                                                    snap HOUSEHOLD_2   
state_id store_id date       year month weekday                        
CA       CA_1     2011-01-29 2011 1     Saturday   False         181   
                  2011-01-30 2011 1     Sunday     False         170   
                  2011-01-31 2011 1     Monday     False         114   
                  2011-02-01 2011 2     Tuesday     True         123   
                  2011-02-02 2011 2     Wednesday   True         135   

                                                                         \
                                                                HOBBIES   
                                                  HOUSEHOLD_1 HOBBIES_2   
state_id store_id date       year month weekday                           
CA       CA_1     2011-01-29 2011 1     Saturday          361        28   
                  2011-01-30 2011 1     Sunday            350         9   
                  2011-01-31 2011 1     Monday            279         6   
                  2011-02-01 2011 2     Tuesday           278         9   
                  2011-02-02 2011 2     Wednesday         195         5   

                                                                             \
                                                              FOODS           
                                                  HOBBIES_1 FOODS_3 FOODS_2   
state_id store_id date       year month weekday                               
CA       CA_1     2011-01-29 2011 1     Saturday        528    2268     674   
                  2011-01-30 2011 1     Sunday          489    2198     655   
                  2011-01-31 2011 1     Monday          409    1398     396   
                  2011-02-01 2011 2     Tuesday         383    1607     476   
                  2011-02-02 2011 2     Wednesday       263    1496     354   

                                                            events            \
                                                              type             
                                                  FOODS_1 sporting religious   
state_id store_id date       year month weekday                                
CA       CA_1     2011-01-29 2011 1     Saturday      297    False     False   
                  2011-01-30 2011 1     Sunday        284    False     False   
                  2011-01-31 2011 1     Monday        214    False     False   
                  2011-02-01 2011 2     Tuesday       175    False     False   
                  2011-02-02 2011 2     Wednesday     182    False     False   

                                                   ...                  \
                                                   ...            name   
                                                   ... independenceday   
state_id store_id date       year month weekday    ...                   
CA       CA_1     2011-01-29 2011 1     Saturday   ...           False   
                  2011-01-30 2011 1     Sunday     ...           False   
                  2011-01-31 2011 1     Monday     ...           False   
                  2011-02-01 2011 2     Tuesday    ...           False   
                  2011-02-02 2011 2     Wednesday  ...           False   

                                                                          \
                                                                           
                                                  halloween father's day   
state_id store_id date       year month weekday                            
CA       CA_1     2011-01-29 2011 1     Saturday      False        False   
                  2011-01-30 2011 1     Sunday        False        False   
                  2011-01-31 2011 1     Monday        False        False   
                  2011-02-01 2011 2     Tuesday       False        False   
                  2

In [9]:
sale_viz.sales_by_store_and_state()

In [10]:
sale_viz.sales_per_category()

Foods_1 sells the most by far with more sales in Summer than Winter.

Hobbies_2 has a peak every October.

Household_1 has a trough every Nov/Dec.

Household_2 is clearly seasonal with fewer items sold in winter.

In [11]:
sale_viz.sales_per_department_and_store()

## To explore

- how are sales on different week days
- compare sales around holidays (on the day and in the run up)
  - does the type of holiday make a difference
  - does the name of holiday make a difference
- sales on snap days
  - this is very nice (https://www.kaggle.com/headsortails/back-to-predict-the-future-interactive-m5-eda/report)